internal/service/storage/storage_backup.go

```diff
@@ -0,0 +1,227 @@
+	resp.TypeName = req.ProviderTypeName + "_storage_backup"
+}
+
+// Configure adds the provider configured client to the resource.
+func (r *storageBackupResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	r.client, resp.Diagnostics = utils.GetClientFromProviderData(req.ProviderData)
+}
+
+type storageBackupModel struct {
+	SourceStorage types.String `tfsdk:"source_storage"`
+	Title         types.String `tfsdk:"title"`
+	ID            types.String `tfsdk:"id"`
+	CreatedAt     types.String `tfsdk:"created_at"`
+}
+
+func (r *storageBackupResource) Schema(_ context.Context, _ resource.SchemaRequest, resp *resource.SchemaResponse) {
+	resp.Schema = schema.Schema{
+		Description: "Manages an on-demand storage backup.",
+		Attributes: map[string]schema.Attribute{
+			"source_storage": schema.StringAttribute{
+				Required:    true,
+				Description: "The UUID of the storage to back up.",
+			},
+			"title": schema.StringAttribute{
+				Required:    true,
+				Description: "Title of the backup.",
+			},
+			"id": schema.StringAttribute{
+				Computed:    true,
+				Description: "ID of the created backup.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+			"created_at": schema.StringAttribute{
+				Computed:    true,
+				Description: "Timestamp of the backup creation.",
+				PlanModifiers: []planmodifier.String{
+					stringplanmodifier.UseStateForUnknown(),
+				},
+			},
+		},
+	}
+}
+
+func (r *storageBackupResource) Create(ctx context.Context, req resource.CreateRequest, resp *resource.CreateResponse) {
+	var data storageBackupModel
+	resp.Diagnostics.Append(req.Plan.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	// Create a backup
+	backupDetails, err := r.client.CreateBackup(ctx, &request.CreateBackupRequest{
+		UUID:  data.SourceStorage.ValueString(),
+		Title: data.Title.ValueString(),
+	})
+	if err != nil {
+		resp.Diagnostics.AddError("Failed to create storage backup", utils.ErrorDiagnosticDetail(err))
+		return
+	}
+
+	// Wait for backup to finish
+	_, err = r.client.WaitForStorageState(ctx, &request.WaitForStorageStateRequest{
+		UUID:         data.SourceStorage.ValueString(),
+		DesiredState: upcloud.StorageStateOnline,
+	})
+	if err != nil {
+		resp.Diagnostics.AddError(
+			"Storage did not reach online state after backup process",
+			utils.ErrorDiagnosticDetail(err),
+		)
+		return
+	}
+
+	data.CreatedAt = types.StringValue(backupDetails.Created.Format(time.RFC3339))
+	data.ID = types.StringValue(backupDetails.UUID)
+	resp.Diagnostics.Append(resp.State.Set(ctx, &data)...)
+}
+
+func (r *storageBackupResource) Read(ctx context.Context, req resource.ReadRequest, resp *resource.ReadResponse) {
+	var data storageBackupModel
+	resp.Diagnostics.Append(req.State.Get(ctx, &data)...)
+
+	if resp.Diagnostics.HasError() {
+		return
+	}
+
+	if data.SourceStorage.ValueString() == "" {
```

Comment:
Should this be `ID` instead of `SourceStorage`? I think the backup can be used also after the source storage has been removed 🤔 

```suggestion
	if data.ID.ValueString() == "" {
```

Reply:
I implemented this change and I got this, which makes sense. I am not sure if it is usable from the point of view of terraform, like how these backups would be used later on. At least, we keep them just in case.
![Screenshot from 2025-04-22 19-38-50](https://github.com/user-attachments/assets/f35134ac-e14e-4590-a483-9e9a023c307e)